R/utils-stats.R

```diff
@@ -1,31 +1,49 @@
 medianCI <- function(x, conf.level=.95) {
-  w <- try(wilcox.test(x, conf.int=T, conf.level=conf.level))
-  if (veupathUtils::is.error(w)) {
-    ci <- list('lowerBound'=jsonlite::unbox(NA), 'upperBound'=jsonlite::unbox(NA), 'error'=jsonlite::unbox(w[1]))
+  x <- x[order(x)]
+  n <- length(x)
+  z <- qnorm(p=(1-conf.level)/2, lower.tail=FALSE)
+  j <- ceiling(.5*n - z*sqrt(n*.25))
+  k <- ceiling(.5*n + z*sqrt(n*.25))
+
+  if (j < 1 || k > length(x)) {
+    ci <- list('lowerBound'=jsonlite::unbox(NA), 'upperBound'=jsonlite::unbox(NA), 'error'=jsonlite::unbox("Failed to determine confidence interval for median."))
   } else {
-    ci <- list('lowerBound'=jsonlite::unbox(w$conf.int[1]), 'upperBound'=jsonlite::unbox(w$conf.int[2]), 'error'=jsonlite::unbox(""))
+    ci <- list('lowerBound'=jsonlite::unbox(x[j]), 'upperBound'=jsonlite::unbox(x[k]), 'error'=jsonlite::unbox(""))
   }
 
   return(list(ci))
 }
 
 meanCI <- function(x, conf.level=.95) {
-  t <- try(t.test(x, conf.int=T, conf.level=conf.level))
-  if (veupathUtils::is.error(t)) {
-    ci <- list('lowerBound'=jsonlite::unbox(NA), 'upperBound'=jsonlite::unbox(NA), 'error'=jsonlite::unbox(t[1]))
+  sd <- sd(x)
+  n <- length(x)
+  z <- qnorm(p=(1-conf.level)/2, lower.tail=FALSE)
+  lowerBound <- mean(x) - z*sd/sqrt(n)
+  upperBound <- mean(x) + z*sd/sqrt(n)
+
+  if (any(is.na(c(lowerBound, upperBound)))) {
+    ci <- list('lowerBound'=jsonlite::unbox(NA), 'upperBound'=jsonlite::unbox(NA), 'error'=jsonlite::unbox("Failed to determine confidence interval for mean."))
   } else {
-    ci <- list('lowerBound'=jsonlite::unbox(t$conf.int[1]), 'upperBound'=jsonlite::unbox(t$conf.int[2]), 'error'=jsonlite::unbox(""))
+    ci <- list('lowerBound'=jsonlite::unbox(lowerBound), 'upperBound'=jsonlite::unbox(upperBound), 'error'=jsonlite::unbox(""))
   }
+  
 
   return(list(ci))
 }
 
 proportionCI <- function(numerator, denominator, conf.level=.95) {
-  b <- try(binom.test(numerator, denominator, conf.level=conf.level))
-  if (veupathUtils::is.error(b)) {
-    ci <- list('lowerBound'=jsonlite::unbox(NA), 'upperBound'=jsonlite::unbox(NA), 'error'=jsonlite::unbox(b[1]))
+  p <- numerator/denominator
+  n <- denominator
+  z <- qnorm(p=(1-conf.level)/2, lower.tail=FALSE)
+  lowerBound <- p - z*sqrt(p*(1-p)/n)
+  lowerBound <- ifelse(lowerBound < 0, 0, lowerBound)
+  upperBound <- p + z*sqrt(p*(1-p)/n)
+  upperBound <- ifelse(upperBound > 1, 1, upperBound)
```

Comment:
standard NA check - if upperBound/lowerBound is NA, will `ifelse` return NA here? I was reading about it but it wasn't totally clear.

Reply:
looks like your tests below cover this case. Nvm

Reply:
something like NA < 0 should return NA, and something like numeric() < 0 should return logical(). both should be caught by the if condition immediately below.